Add verifier refactor docs (#3)
veraison/docs: architecture/verifier/README.md

# Verifier Refactor Changes Overview

## Configuration

Configuration for each Veraison component is encapsulated by a `ParamStore`
that contains parameter definitions. A parameter definition specifies the type
of the parameter (used in validation), and a path where it can be found within
configuration.

Configuration for a Veraison deployment is handled by a `Config` that gathers
`ParamStore`s for components used in that deployment. The `Config`uses
[Viper](https://github.com/spf13/viper) to gather configuration from various
sources and populate `ParamStore`s associated with it. Parameters are populated
based on their paths,  and the same value can be used to populate parameters
inside multiple stores, if they share a path.

<p/>
<p align="center">
  <img src="./veraison-config.png"/>
</p>

> `ParamStore`s are created, populated with parameter definitions for their

Reviewer: `ParamStore`s: We need to correct some indentation here, as it does not come up nicely in the regular text ?

Author: Sorry, what do you mean by "indentation" here? If you're talking about about the gap between the "e" and the "s" in "`ParamStore`s" then there is not much we can do about that, as this is how git hub renders this. The alternatives would be
- `ParamStores` -- this implies that the plural form itself an identifier and may lead to confusion.
-  `ParamStore`'s -- the `'` is sometimes used in lieu of formatting to indicate pluralisation of a singular identifier (rather than the plural form itself being an identifier) in plain text. However, given that formatting *is* available, I would argue the apostrophe is not desirable, as it is already overloaded to indicate possessive form and the contraction of "is".
- `ParamStore` instances -- avoids needing to pluralize the identifier at the expense of making the sentence longer and more complex. I would argue this would have a net negative effect on comprehension.

IMO, the current structure is the least bad of the options. However, this is stylistic, and I would be happy to change it to one of the above options?

Reviewer: Yes, I meant a rendering issue. Does not come up nicely. Let me ponder over but I agree, we may have to leave it like this, if we do not have other simple option to fix it!

> `ParamStore`s are created, populated with parameter definitions for their

Reviewer: Also one more point I noticed is that here in the text we use ParamStore, however in the figures we use ParmStore, so just a Nit, we should be consistent!

Author: oh, that's a typo in the figure; fill fix.

> `ParamStore`s are created, populated with parameter definitions for their

Reviewer: Also could you please clarify or break the statement at line 22, I am having little difficulty understanding the flow.
`ParamStore`s are created, populated with parameter definitions for their respective components (by whom ?)

Author: done.

respective component, and then passed onto the `Config` that assumes ownership

> of them, and populates them from external sources. This happens on deployment

Reviewer: So does line 24 imply that that ParamStore is re-populated?

Author: No, the `Config` populates them with actual values. The text above take about populating them with parameter definition.

I can see how this is confusing. Will reword.

Reviewer: Thanks!

start. Once the `Config` finishes processing external sources, deployment
initialization proceeds by initializing the individual components. The

> `ParamStore`s are passed to the components' `Init()` methods, whereupon the

Reviewer: Same as line 22

Author: See above.

`Config` relinquishes ownership over them (i.e. they would no longer be
updated). The values in the stores are used by the `Init()` methods to
initialize the components. `ParamStores` are not used beyond that.

There is currently no provision for continuous monitoring of external sources
and dynamic update mid-execution.

## Verifier

The Verifier receives an attestation token and returns an attestation result

> indicating whither then token is well structured and has been verified against

Reviewer: whether that

provisioned endorsements. Most of the heavy lifting is performed by the
Veraison Trusted Services (VTS) component. It can be executed within the same
process as the verifier (for integrated deployment) or in a separate process
(and possibly, on a separate node), in order to isolate high-trust services.

VTS verifies the token against the trust anchor, extracts the evidence, fetches
associated endorsements, and initializes an attestation result, populating the
Trust Vector by appraising the evidence against the endorsements in an
attestation scheme specific manner.

The resulting attestation is returned to the verifier. If a policy has been
registered as part of the attestation scheme, the verifier will  apply it to
the attestation, allowing for the Trust Vector to be updated in a custom
manner.

<p/>
<p align="center">
  <img src="./verifier-architecture.png"/>
</p>

## VTS

Veraison Trusted Services (in the context of verification) consist of

- Schemes that encapsulate the knowledge of how a particular token should be

>   parsed in order to extract evidence, how the token maybe validated against a

Reviewer 2: ```suggestion
  parsed in order to extract evidence, how the token may be validated against a
```

  trust anchor, and how evidence should be evaluated against endorsements in
  order to populate the Trust Vector.
- A trust anchor store that maintains provisioned trust anchors (such as keys
  and certificates) used to validate received tokens.
- An endorsement store that maintains endorsements provisioned from the supply

>   chain (OEM, ISV, etc). Endorsements may be "gold values" again which claims

Reviewer 2: ```suggestion
  chain (OEM, ISV, etc). Endorsements may be "golden values" again which claims
```

>   chain (OEM, ISV, etc). Endorsements may be "gold values" again which claims

Author: yup.

  from the evidences can be verified, or additional claims that are associated
  with, but are not part of the token (e.g. certification information).

> On the provisioning path, VTS interface exposes a single method that processes

Reviewer 2: ```suggestion
On the verification path, VTS interface exposes a single method that processes
```

`ParamStore`s for components used in that deployment. The `Config`uses
[Viper](https://github.com/spf13/viper) to gather configuration from various
sources and populate `ParamStore`s associated with it. Parameters are populated
based on their paths,  and the same value can be used to populate parameters
inside multiple stores, if they share a path.

<p/>
<p align="center">
  <img src="./veraison-config.png"/>
</p>

`ParamStore`s are created and initialized with parameter definitions for their
respective component. Most components will have associated helper functions to
do that. The `ParamStore`s are then passed onto the `Config` that assumes
ownership of them, and populates them with values read from external sources.
The values are validated against the definitions. This happens on deployment
start. Once the `Config` finishes processing external sources, deployment
initialization proceeds by initializing the individual components. The
`ParamStore`s are passed to the components' `Init()` methods, whereupon the
`Config` relinquishes ownership over them (i.e. they would no longer be
updated). The values in the stores are used by the `Init()` methods to

> initialize the components. `ParamStores` are not used beyond that.

Reviewer: On line 32 we have used the term `ParamStores` but I admit it does not look bad either ....
